pkg/controller/rollout/rollout_bluegreen.go

```diff
@@ -0,0 +1,487 @@
+		klog.Infof("rollout(%s/%s) batchRelease status(%s) is not ready, and wait a moment", c.Rollout.Namespace, c.Rollout.Name, util.DumpJSON(br.Status))
+		return false, nil
+	}
+	m.recorder.Eventf(c.Rollout, corev1.EventTypeNormal, "Progressing", fmt.Sprintf("upgrade step(%d) canary pods with new versions done", blueGreenStatus.CurrentStepIndex))
+	klog.Infof("rollout(%s/%s) batch(%s) state(%s), and success",
+		c.Rollout.Namespace, c.Rollout.Name, util.DumpJSON(br.Status), br.Status.CanaryStatus.CurrentBatchState)
+	// set the latest PodTemplateHash to selector the latest pods.
+	blueGreenStatus.PodTemplateHash = c.Workload.PodTemplateHash
+	return true, nil
+}
+
+func (m *blueGreenReleaseManager) doCanaryMetricsAnalysis(c *RolloutContext) (bool, error) {
+	// todo
+	return true, nil
+}
+
+func (m *blueGreenReleaseManager) doCanaryPaused(c *RolloutContext) (bool, error) {
+	blueGreenStatus := c.NewStatus.BlueGreenStatus
+	currentStep := c.Rollout.Spec.Strategy.BlueGreen.Steps[blueGreenStatus.CurrentStepIndex-1]
+	steps := len(c.Rollout.Spec.Strategy.BlueGreen.Steps)
+	cond := util.GetRolloutCondition(*c.NewStatus, v1beta1.RolloutConditionProgressing)
+	// need manual confirmation
+	if currentStep.Pause.Duration == nil {
+		klog.Infof("rollout(%s/%s) don't set pause duration, and need manual confirmation", c.Rollout.Namespace, c.Rollout.Name)
+		cond.Message = fmt.Sprintf("Rollout is in step(%d/%d), and you need manually confirm to enter the next step", blueGreenStatus.CurrentStepIndex, steps)
+		c.NewStatus.Message = cond.Message
+		return false, nil
+	}
+	cond.Message = fmt.Sprintf("Rollout is in step(%d/%d), and wait duration(%d seconds) to enter the next step", blueGreenStatus.CurrentStepIndex, steps, *currentStep.Pause.Duration)
+	c.NewStatus.Message = cond.Message
+	// wait duration time, then go to next step
+	duration := time.Second * time.Duration(*currentStep.Pause.Duration)
+	expectedTime := blueGreenStatus.LastUpdateTime.Add(duration)
+	if expectedTime.Before(time.Now()) {
+		klog.Infof("rollout(%s/%s) canary step(%d) paused duration(%d seconds), and go to the next step",
+			c.Rollout.Namespace, c.Rollout.Name, blueGreenStatus.CurrentStepIndex, *currentStep.Pause.Duration)
+		return true, nil
+	}
+	c.RecheckTime = &expectedTime
+	return false, nil
+}
+
+func (m *blueGreenReleaseManager) doCanaryJump(c *RolloutContext) (jumped bool) {
+	bluegreenStatus := c.NewStatus.BlueGreenStatus
+	nextIndex := bluegreenStatus.NextStepIndex
+	/*
+		we set the CurrentStepIndex same as NextStepIndex to prevent currentStepIndex from out of range
+		for example, if we had a rollout with 4 steps and CurrentStepIndex was 2
+		then, the user removed 3 steps from the plan, we can calculate NextStepIndex is 1 correctly,
+		but CurrentStepIndex remains 2, which could cause out of range.
+	*/
+	resetCurrentIndex := false
+	if int(bluegreenStatus.CurrentStepIndex) > len(c.Rollout.Spec.Strategy.BlueGreen.Steps) {
```

Comment:
is it possible to forbid adding or removing steps during rollout so that the code can be more simpler?

Reply:
done